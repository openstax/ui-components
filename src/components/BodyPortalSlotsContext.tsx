import React from 'react';

export const BodyPortalSlotsContext = React.createContext<string[]>([]);
