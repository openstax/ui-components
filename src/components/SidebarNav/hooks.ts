import React from "react";
import { useMatchMediaQuery } from "../../../src/hooks";
import { breakpoints } from "../../../src/theme";

export const useSidebarNavProps = ({
  mobileBreakpoint = `${breakpoints.mobileNavBreak}em`,
  ...props
}: {
  isMobile?: boolean;
  mobileBreakpoint?: string;
}) => {
  const mobileQueryMatches = useMatchMediaQuery(
    `(max-width: ${mobileBreakpoint})`,
  );
  const isMobile = props.isMobile ?? mobileQueryMatches;
  const [navIsCollapsed, setNavIsCollapsed] = React.useState(isMobile);

  return { isMobile, navIsCollapsed, setNavIsCollapsed };
};

export const useNavAnimation = () => {
  // Transition CSS rules won't work with the BodyPortal becase the
  // nodes get reinserted, so use a class name for @keyframes instead.
  // There is an awkward empty state here - we need to distinguish
  // "idle after an interaction" vs. the initial page load.
  const [navAnimation, setNavAnimation] = React.useState<
    "expanding" | "collapsing" | "idle" | ""
  >("");

  React.useEffect(() => {
    if (!navAnimation || navAnimation === "idle") {
      return;
    }

    const idleCallback = setTimeout(() => setNavAnimation("idle"), 300);
    return () => clearTimeout(idleCallback);
  }, [navAnimation, setNavAnimation]);

  return { navAnimation, setNavAnimation };
};
