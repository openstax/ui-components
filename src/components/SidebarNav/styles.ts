import { colors, zIndex } from "../../../src/theme";
import styled, { css } from "styled-components";

export const collapsedWidth = "5.6rem";
export const expandedWidth = "24rem";
export const navStyles = css`
  --collapsed-width: ${collapsedWidth};
  --expanded-width: ${expandedWidth};
  width: var(--expanded-width);
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  position: relative;
  box-shadow: -0.2rem 0 0.4rem rgba(0, 0, 0, 0.1) inset;
  background: ${colors.palette.neutralBright};
  color: ${colors.palette.neutralThin};
  z-index: ${zIndex.navbar - 1};

  @media (max-width: 15em) {
    --expanded-width: 100vw;
    width: 100vw;
  }

  &.collapsed {
    width: 5.6rem;
  }

  &.mobile {
    position: fixed;
    height: 100%;
  }

  &.mobile + nav,
  &.mobile + nav + main {
    margin-left: ${collapsedWidth};
  }

  &.mobile ~ main::before,
  &.mobile ~ [data-backdrop-target]::before {
    content: "";
    background: none;
    top: 0;
    left: 0;
    width: 0;
    height: 0;
    position: fixed;
    opacity: 0;
    transition: opacity 300ms ease-in-out;
    z-index: ${zIndex.navbar + 1};
  }

  &.mobile:not(.collapsed),
  &.mobile.collapsing {
    z-index: ${zIndex.sidebar};

    & ~ main::before,
    & ~ [data-backdrop-target]::before {
      background: rgba(0 0 0 / 0.7);
      opacity: 1;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      width: 100%;
      height: 100%;
      z-index: ${zIndex.sidebar - 1};
    }
  }

  &.mobile.collapsing {
    & ~ main::before,
    & ~ [data-backdrop-target]::before {
      opacity: 0;
    }
  }

  @keyframes expandSidebarNav {
    from {
      width: var(--collapsed-width);
    }
    to {
      width: var(--expanded-width);
    }
  }
  @keyframes collapseSidebarNav {
    from {
      width: var(--expanded-width);
    }
    to {
      width: var(--collapsed-width);
    }
  }
  &.expanding {
    animation: expandSidebarNav 300ms forwards;
  }
  &.collapsing {
    animation: collapseSidebarNav 300ms forwards;
  }
`;

export const NavHeader = styled.header`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
`;

export const NavBody = styled.div`
  height: 100%;
  overflow-y: auto;
`;

export const NavFooter = styled.footer`
  button {
    width: 100%;
  }
`;

export const ToggleButton = styled.button`
  position: absolute;
  right: 0;
  top: 1.8rem;
  width: 1.6rem;
  height: 2.4rem;
  background: #fff;
  padding: 0;
  border: 0;
  cursor: pointer;
  border: 0.1rem solid #959595;
  border-right: 0;
  border-radius: 0.4rem 0 0 0.4rem;
  z-index: 1;
`;
