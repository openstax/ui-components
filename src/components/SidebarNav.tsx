import React from "react";
import classNames from "classnames";
import styled from "styled-components";
import { LeftArrow } from "./svgs/LeftArrow";
import { RightArrow } from "./svgs/RightArrow";
import { breakpoints, colors, zIndex } from "../theme";
import { useMatchMediaQuery } from "../hooks";
import { FocusScope } from "react-aria";

const collapsedWidth = "5.6rem";
const expandedWidth = "24rem";

const Nav = styled.nav`
  --collapsed-width: ${collapsedWidth};
  --expanded-width: ${expandedWidth};
  width: var(--expanded-width);
  transition: width 300ms ease-in-out;
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  position: relative;
  box-shadow: -0.2rem 0 0.4rem rgba(0, 0, 0, 0.1) inset;
  background: ${colors.palette.neutralBright};
  color: ${colors.palette.neutralThin};
  z-index: ${zIndex.navbar - 1};

  @media (max-width: 15em) {
    --expanded-width: 100vw;
    width: 100vw;
  }

  &.collapsed {
    width: 5.6rem;
  }

  &.mobile {
    position: fixed;
    height: 100%;
  }

  &.mobile ~ main::before {
    content: "";
    background: none;
    top: 0;
    left: 0;
    width: 0;
    height: 0;
    position: fixed;
    opacity: 0;
    transition: opacity 300ms ease-in-out;
  }

  &.mobile[aria-expanded="true"] ~ main::before,
  &.mobile[aria-expanded="true"] ~ [data-backdrop-target]::before {
    background: rgba(0 0 0 / .7);
    opacity: 1;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    width: 100%;
    height: 100%;
  }
`;

const NavHeader = styled.header`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
`;

const NavBody = styled.div`
  height: 100%;
  overflow-y: auto;
`;

const NavFooter = styled.footer`
  button {
    width: 100%;
  }
`;

const ToggleButton = styled.button`
  position: absolute;
  right: 0;
  top: 1.8rem;
  width: 1.6rem;
  height: 2.4rem;
  background: #fff;
  padding: 0;
  border: 0;
  cursor: pointer;
  border: 0.1rem solid #959595;
  border-right: 0;
  border-radius: 0.4rem 0 0 0.4rem;
`;

type FunctionRender = (_: {
  navIsCollapsed: boolean;
  setNavIsCollapsed: (_: boolean) => void;
}) => React.ReactNode;

export const SidebarNav = styled(
  ({
    navHeader,
    navFooter,
    children,
    className,
    mobileBreakpoint = `${breakpoints.mobileBreak}em`,
    ...props
  }: React.HTMLAttributes<HTMLElement> & {
    navHeader?: React.ReactNode | FunctionRender;
    navFooter?: React.ReactNode | FunctionRender;
    children: React.ReactNode | FunctionRender;
    isMobile?: boolean;
    mobileBreakpoint?: string;
  }) => {
    const mobileQueryMatches = useMatchMediaQuery(
      `(max-width: ${mobileBreakpoint})`,
    );
    const isMobile = props.isMobile ?? mobileQueryMatches;

    const [navIsCollapsed, setNavIsCollapsed] = React.useState(isMobile);
    const toggleButtonRef = React.useRef<HTMLButtonElement>(null);
    const sidebarNavRef = React.useRef<HTMLDivElement>(null);

    React.useLayoutEffect(() => {
      setNavIsCollapsed(isMobile);
    }, [isMobile]);

    React.useEffect(() => {
      if (!isMobile || navIsCollapsed) {
        return;
      }

      const handleOutsideEvent = (event: any) => {
        if (
          isMobile &&
          !navIsCollapsed &&
          sidebarNavRef.current &&
          !sidebarNavRef.current.contains(event.target)
        ) {
          setNavIsCollapsed(true);
        }
      };

      const handleKeyPress = (event: KeyboardEvent) => {
        if (isMobile && !navIsCollapsed && event.key === "Escape") {
          setNavIsCollapsed(true);
        }
      };

      document.addEventListener("mousedown", handleOutsideEvent);
      document.addEventListener("touchstart", handleOutsideEvent);
      document.addEventListener("keydown", handleKeyPress);

      return () => {
        document.removeEventListener("click", handleOutsideEvent);
        document.removeEventListener("touchstart", handleOutsideEvent);
        document.removeEventListener("keydown", handleKeyPress);
      };
    }, [isMobile, navIsCollapsed, setNavIsCollapsed, sidebarNavRef]);

    const functionRenderArguments = {
      navIsCollapsed,
      setNavIsCollapsed,
    };

    return (
      <FocusScope contain={isMobile && !navIsCollapsed}>
        <Nav
          ref={sidebarNavRef}
          aria-expanded={!navIsCollapsed}
          data-testid="sidebarnav"
          className={classNames(className, {
            collapsed: navIsCollapsed,
            mobile: isMobile,
          })}
          {...props}
        >
          <ToggleButton
            ref={toggleButtonRef}
            data-testid="sidebarnav-toggle"
            className={classNames({ collapsed: navIsCollapsed })}
            onClick={() => setNavIsCollapsed(!navIsCollapsed)}
            aria-label={
              navIsCollapsed ? "Expand navigation" : "Collapse navigation"
            }
          >
            {navIsCollapsed ? <RightArrow /> : <LeftArrow />}
          </ToggleButton>
          {navHeader ? (
            <NavHeader>
              {typeof navHeader === "function"
                ? navHeader(functionRenderArguments)
                : navHeader}
            </NavHeader>
          ) : null}
          <NavBody>
            {typeof children === "function"
              ? children(functionRenderArguments)
              : children}
          </NavBody>
          {navFooter ? (
            <NavFooter>
              {typeof navFooter === "function"
                ? navFooter(functionRenderArguments)
                : navFooter}
            </NavFooter>
          ) : null}
        </Nav>
      </FocusScope>
    );
  },
)``;

export const SidebarNavStyles = {
  Nav,
  NavHeader,
  NavBody,
  NavFooter,
  ToggleButton,
  expandedWidth,
  collapsedWidth,
};
