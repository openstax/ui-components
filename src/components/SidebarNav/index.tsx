import React from "react";
import classNames from "classnames";
import styled from "styled-components";
import { LeftArrow } from "../svgs/LeftArrow";
import { RightArrow } from "../svgs/RightArrow";
import { FocusScope } from "react-aria";
import { BodyPortal } from "../BodyPortal";

import {
  NavBody,
  NavFooter,
  NavHeader,
  ToggleButton,
  expandedWidth,
  collapsedWidth,
  navStyles,
} from "./styles";
import { useNavAnimation, useSidebarNavProps } from "./hooks";

type FunctionRender = (_: {
  navIsCollapsed: boolean;
  setNavIsCollapsed: (_: boolean) => void;
  isMobile: boolean;
}) => React.ReactNode;

interface SidebarNavSharedProps {
  id?: string;
  navHeader?: React.ReactNode | FunctionRender;
  navFooter?: React.ReactNode | FunctionRender;
  children: React.ReactNode | FunctionRender;
  mobileBreakpoint?: string;
  isMobile?: boolean;
  className?: string;
  ariaLabel?: string;
}

export const SidebarNavBase = ({
  sidebarNavRef,
  navHeader,
  navFooter,
  children,
  navIsCollapsed,
  setNavIsCollapsed,
  navAnimation,
  isMobile,
}: SidebarNavSharedProps & {
  sidebarNavRef?: React.MutableRefObject<HTMLElement | null>;
  navIsCollapsed: boolean;
  setNavIsCollapsed: React.Dispatch<boolean>;
  navAnimation?: string;
  isMobile: boolean;
}) => {
  const toggleButtonRef = React.useRef<HTMLButtonElement>(null);

  React.useLayoutEffect(() => {
    setNavIsCollapsed(isMobile);
  }, [isMobile]); // eslint-disable-line react-hooks/exhaustive-deps

  React.useEffect(() => {
    if (!isMobile || navIsCollapsed) {
      return;
    }

    const handleOutsideEvent = (event: any) => {
      if (
        isMobile &&
        !navIsCollapsed &&
        sidebarNavRef?.current &&
        !sidebarNavRef.current.contains(event.target) &&
        document.body.contains(event.target)
      ) {
        setNavIsCollapsed(true);
      }
    };

    const handleKeyPress = (event: KeyboardEvent) => {
      if (isMobile && !navIsCollapsed && event.key === "Escape") {
        setNavIsCollapsed(true);
      }
    };

    document.addEventListener("click", handleOutsideEvent);
    document.addEventListener("touchend", handleOutsideEvent);
    document.addEventListener("keydown", handleKeyPress);

    return () => {
      document.removeEventListener("click", handleOutsideEvent);
      document.removeEventListener("touchend", handleOutsideEvent);
      document.removeEventListener("keydown", handleKeyPress);
    };
  }, [isMobile, navIsCollapsed, setNavIsCollapsed, sidebarNavRef]);

  const functionRenderArguments = {
    navIsCollapsed,
    setNavIsCollapsed,
    isMobile,
  };

  React.useEffect(() => {
    if (navAnimation === "idle") {
      (toggleButtonRef.current as HTMLButtonElement).focus();
    }
  }, [navAnimation]);

  const navBodyRef = React.useRef<HTMLDivElement>(null);
  const [scrollPosition, setScrollPosition] = React.useState(0);

  requestAnimationFrame(() => {
    if (navBodyRef.current) {
      navBodyRef.current.scrollTop = scrollPosition;
    }
  });

  return (
    <FocusScope contain={isMobile && !navIsCollapsed}>
      <ToggleButton
        aria-expanded={!navIsCollapsed}
        ref={toggleButtonRef}
        data-testid="sidebarnav-toggle"
        className={classNames({ collapsed: navIsCollapsed })}
        onClick={(e) => {
          setNavIsCollapsed(!navIsCollapsed);
          e.stopPropagation();
        }}
        aria-label={
          navIsCollapsed ? "Expand navigation" : "Collapse navigation"
        }
      >
        {navIsCollapsed ? <RightArrow /> : <LeftArrow />}
      </ToggleButton>
      {navHeader ? (
        <NavHeader>
          {typeof navHeader === "function"
            ? navHeader(functionRenderArguments)
            : navHeader}
        </NavHeader>
      ) : null}
      <NavBody
        data-testid="nav-body"
        ref={navBodyRef}
        onScroll={(e) =>
          setScrollPosition((e.target as HTMLDivElement).scrollTop)
        }
      >
        {typeof children === "function"
          ? children(functionRenderArguments)
          : children}
      </NavBody>
      {navFooter ? (
        <NavFooter>
          {typeof navFooter === "function"
            ? navFooter(functionRenderArguments)
            : navFooter}
        </NavFooter>
      ) : null}
    </FocusScope>
  );
};

export const SidebarNav = styled(
  ({ className, id, ariaLabel, ...props }: SidebarNavSharedProps) => {
    const { isMobile, navIsCollapsed, setNavIsCollapsed } =
      useSidebarNavProps(props);
    const sidebarNavRef = React.useRef<HTMLElement>(null);
    const { navAnimation, setNavAnimation } = useNavAnimation();

    const handleSetNavIsCollapsed = (value: boolean) => {
      if (value !== navIsCollapsed) {
        setNavAnimation(value ? "collapsing" : "expanding");
      }

      setNavIsCollapsed(value);
    };

    return (
      <nav
        id={id}
        ref={sidebarNavRef}
        data-testid="sidebarnav"
        aria-label={ariaLabel}
        className={classNames(className, {
          collapsed: navIsCollapsed,
          mobile: isMobile,
          collapsing: navAnimation === "collapsing",
          expanding: navAnimation === "expanding",
        })}
      >
        <SidebarNavBase
          {...props}
          sidebarNavRef={sidebarNavRef}
          navIsCollapsed={navIsCollapsed}
          setNavIsCollapsed={handleSetNavIsCollapsed}
          isMobile={isMobile}
        >
          {props.children}
        </SidebarNavBase>
      </nav>
    );
  },
)`
  ${navStyles}
`;

export const BodyPortalSidebarNav = styled(
  ({ className, id, ariaLabel, ...props }: SidebarNavSharedProps) => {
    const { isMobile, navIsCollapsed, setNavIsCollapsed } =
      useSidebarNavProps(props);

    const ref = React.useRef<HTMLElement>(document.createElement("NAV"));
    const { navAnimation, setNavAnimation } = useNavAnimation();

    const handleSetNavIsCollapsed = (value: boolean) => {
      if (value !== navIsCollapsed) {
        setNavAnimation(value ? "collapsing" : "expanding");
      }

      setNavIsCollapsed(value);
    };

    return (
      <BodyPortal
        ref={ref}
        id={id}
        tagName="nav"
        slot="sidebar"
        data-testid="sidebarnav"
        aria-label={ariaLabel}
        className={classNames(className, {
          collapsed: navIsCollapsed,
          mobile: isMobile,
          collapsing: navAnimation === "collapsing",
          expanding: navAnimation === "expanding",
        })}
      >
        <SidebarNavBase
          {...props}
          navIsCollapsed={navIsCollapsed}
          setNavIsCollapsed={handleSetNavIsCollapsed}
          sidebarNavRef={ref}
          navAnimation={navAnimation}
          isMobile={isMobile}
        />
      </BodyPortal>
    );
  },
)`
  ${navStyles}
`;

export const SidebarNavStyles = {
  NavHeader,
  NavBody,
  NavFooter,
  ToggleButton,
  expandedWidth,
  collapsedWidth
};
