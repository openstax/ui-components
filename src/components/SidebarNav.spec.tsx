import {
  render,
  screen,
  fireEvent,
  act,
  waitFor,
  cleanup,
} from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import { BodyPortalSidebarNav, SidebarNav, SidebarNavBase } from "./SidebarNav";
import "@testing-library/jest-dom";
import { BodyPortalSlotsContext } from "./BodyPortalSlotsContext";

jest.useFakeTimers();

describe("SidebarNav", () => {
  let user: UserEvent;

  beforeAll(() => {
    user = userEvent.setup();
  });

  beforeEach(() => {
    window.matchMedia = jest.fn().mockImplementation(() => {
      return {
        matches: false,
        addListener: jest.fn(),
        removeListener: jest.fn(),
      };
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders header and footer when provided", () => {
    const header = <div>Header</div>;
    const footer = <div>Footer</div>;
    render(
      <SidebarNav
        navHeader={header}
        navFooter={footer}
        mobileBreakpoint="100px"
      >
        Content
      </SidebarNav>,
    );
    expect(screen.getByText("Header")).toBeInTheDocument();
    expect(screen.getByText("Footer")).toBeInTheDocument();
  });

  it("renders header and footer functions", () => {
    const header = <div>Header</div>;
    const footer = <div>Footer</div>;
    render(
      <SidebarNav
        navHeader={() => header}
        navFooter={() => footer}
        mobileBreakpoint="100px"
      >
        Content
      </SidebarNav>,
    );
    expect(screen.getByText("Header")).toBeInTheDocument();
    expect(screen.getByText("Footer")).toBeInTheDocument();
  });

  it("toggles nav collapse on button click", async () => {
    window.matchMedia = jest.fn().mockImplementation(() => {
      return {
        matches: false,
        addListener: jest.fn(),
        removeListener: jest.fn(),
      };
    });

    const component = render(<SidebarNav>Content</SidebarNav>);
    expect(component.asFragment()).toMatchSnapshot();

    expect(screen.getByRole("navigation")).not.toHaveClass("collapsed");
    expect(screen.getByTestId("sidebarnav-toggle")).toHaveAttribute(
      "aria-expanded",
      "true",
    );

    act(() => {
      fireEvent.click(screen.getByTestId("sidebarnav-toggle"));
    });

    expect(screen.getByRole("navigation")).toHaveClass("collapsed");
    expect(screen.getByTestId("sidebarnav-toggle")).toHaveAttribute(
      "aria-expanded",
      "false",
    );
    expect(component.asFragment()).toMatchSnapshot();

    act(() => {
      fireEvent.click(screen.getByTestId("sidebarnav-toggle"));
    });

    expect(screen.getByRole("navigation")).not.toHaveClass("collapsed");
  });

  it("is dismissable on mobile", async () => {
    render(
      <SidebarNav isMobile={true}>
        {({ setNavIsCollapsed }) => (
          <button
            data-testid="close-sidebar"
            onClick={() => setNavIsCollapsed(true)}
          >
            Close
          </button>
        )}
      </SidebarNav>,
    );

    act(() => {
      fireEvent.click(screen.getByTestId("sidebarnav-toggle"));
    });

    act(() => {
      user.type(document.body, "{Escape}");
    });

    await waitFor(() => {
      expect(screen.getByRole("navigation")).toHaveClass("collapsed");
    });
  });

  it("collapses on outside click", async () => {
    render(<SidebarNav isMobile={true}>Content</SidebarNav>);

    act(() => {
      fireEvent.click(screen.getByTestId("sidebarnav-toggle"));
    });

    fireEvent.mouseDown(screen.getByTestId("sidebarnav"));

    await waitFor(() => {
      expect(screen.getByTestId("sidebarnav")).not.toHaveClass("collapsed");
    });

    fireEvent.mouseDown(document);

    await waitFor(() => {
      expect(screen.getByTestId("sidebarnav")).toHaveClass("collapsed");
    });
  });

  it("doesn't collapse on outside click when mobile is false", async () => {
    render(<SidebarNav isMobile={false}>Content</SidebarNav>);

    await waitFor(() => {
      expect(screen.getByTestId("sidebarnav")).not.toHaveClass("collapsed");
    });

    fireEvent.mouseDown(screen.getByTestId("sidebarnav"));

    await waitFor(() => {
      expect(screen.getByTestId("sidebarnav")).not.toHaveClass("collapsed");
    });

    fireEvent.mouseDown(document);

    await waitFor(() => {
      expect(screen.getByTestId("sidebarnav")).not.toHaveClass("collapsed");
    });
  });

  describe("SidebarNavBase", () => {
    it("outside clicks don't set nav to collapsed if the ref is null", () => {
      const setNavCollapsedFn = jest.fn();
      render(
        <SidebarNavBase
          isMobile={true}
          navIsCollapsed={false}
          setNavIsCollapsed={setNavCollapsedFn}
        >
          Content
        </SidebarNavBase>
      );
      // setNavIsCollapsed fires on mount
      expect(setNavCollapsedFn).toHaveBeenCalledTimes(1);
      setNavCollapsedFn.mockReset();
      
      fireEvent.click(document);

      expect(setNavCollapsedFn).not.toHaveBeenCalled();

      cleanup();

      render(
        <SidebarNavBase
          isMobile={true}
          navIsCollapsed={false}
          setNavIsCollapsed={setNavCollapsedFn}
          sidebarNavRef={{ current: null }}
        >
          Content
        </SidebarNavBase>
      );

      setNavCollapsedFn.mockReset();
      fireEvent.click(document);

      expect(setNavCollapsedFn).not.toHaveBeenCalled();
    });
  });

  describe("BodyPortalSidebarNav", () => {
    let root: HTMLElement;

    beforeEach(() => {
      root = document.createElement("main");
      root.id = "root";
      document.body.append(root);
    });

    it("uses a BodyPortal", async () => {
      render(
        <BodyPortalSlotsContext.Provider value={["sidebar", "root"]}>
          <BodyPortalSidebarNav isMobile={false}>
            Sidebar Nav
          </BodyPortalSidebarNav>
          Main
        </BodyPortalSlotsContext.Provider>,
        { container: root },
      );

      expect(document.body).toMatchInlineSnapshot(`
<body>
  <nav
    class="sc-jSMfEi cTSawD"
    data-portal-slot="sidebar"
    data-testid="sidebarnav"
  >
    <span
      data-focus-scope-start="true"
      hidden=""
    />
    <button
      aria-expanded="true"
      aria-label="Collapse navigation"
      class="sc-hKMtZM brWRpF"
      data-testid="sidebarnav-toggle"
    >
      <svg
        fill="none"
        height="10"
        viewBox="0 0 8 10"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.1266 5.38862L5.57702 9.83902C5.79166 10.0537 6.13964 10.0537 6.35426 9.83902L6.87333 9.31995C7.0876 9.10568 7.08801 8.75841 6.87424 8.54363L3.34721 4.99999L6.87424 1.45637C7.08801 1.24159 7.0876 0.89432 6.87333 0.680047L6.35426 0.160979C6.13962 -0.0536598 5.79164 -0.0536598 5.57702 0.160979L1.12662 4.61138C0.911981 4.826 0.911981 5.17398 1.1266 5.38862Z"
          fill="#959595"
        />
      </svg>
    </button>
    <div
      class="sc-gsnTZi hSiqlK"
      data-testid="nav-body"
    >
      Sidebar Nav
    </div>
    <span
      data-focus-scope-end="true"
      hidden=""
    />
  </nav>
  <main
    id="root"
  >
    Main
  </main>
</body>
`);
    });

    it("adds animation classes when opening and closing", () => {
      // Start expanded
      render(
        <BodyPortalSlotsContext.Provider value={["sidebar", "root"]}>
          <BodyPortalSidebarNav isMobile={false}>
            Sidebar Nav
          </BodyPortalSidebarNav>
          Main
        </BodyPortalSlotsContext.Provider>,
        { container: root },
      );

      expect(screen.getByRole("navigation")).not.toHaveClass("collapsing");
      expect(screen.getByRole("navigation")).not.toHaveClass("expanding");

      // collapse
      act(() => {
        fireEvent.click(screen.getByTestId("sidebarnav-toggle"));
      });

      expect(screen.getByRole("navigation")).toHaveClass("collapsing");
      expect(screen.getByRole("navigation")).not.toHaveClass("expanding");

      act(() => {
        jest.advanceTimersByTime(300);
      });

      expect(screen.getByRole("navigation")).not.toHaveClass("collapsing");

      // expand
      act(() => {
        fireEvent.click(screen.getByTestId("sidebarnav-toggle"));
      });

      expect(screen.getByRole("navigation")).not.toHaveClass("collapsing");
      expect(screen.getByRole("navigation")).toHaveClass("expanding");

      act(() => {
        jest.advanceTimersByTime(300);
      });

      expect(screen.getByRole("navigation")).not.toHaveClass("expanding");
    });
  });

  it("handles the onScroll event", async () => {
    render(
      <SidebarNav isMobile={false}>
        {() => <div style={{ height: "200vh" }}>Content</div>}
      </SidebarNav>,
    );

    const navBody = screen.getByTestId("nav-body");

    act(() => {
      fireEvent.scroll(navBody, { target: { scrollTop: 100 } });
    });

    await waitFor(() => {
      expect(navBody.scrollTop).toBe(100);
    });
  });
});
