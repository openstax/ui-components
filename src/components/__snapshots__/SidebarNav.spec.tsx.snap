// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarNav toggles nav collapse on button click 1`] = `
<DocumentFragment>
  <span
    data-focus-scope-start="true"
    hidden=""
  />
  <nav
    aria-expanded="true"
    class="sc-bczRLJ fLqWsF sc-jSMfEi"
    data-testid="sidebarnav"
  >
    <button
      aria-label="Collapse navigation"
      class="sc-eCYdqJ eKnMif"
      data-testid="sidebarnav-toggle"
    >
      <svg
        fill="none"
        height="10"
        viewBox="0 0 8 10"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.1266 5.38862L5.57702 9.83902C5.79166 10.0537 6.13964 10.0537 6.35426 9.83902L6.87333 9.31995C7.0876 9.10568 7.08801 8.75841 6.87424 8.54363L3.34721 4.99999L6.87424 1.45637C7.08801 1.24159 7.0876 0.89432 6.87333 0.680047L6.35426 0.160979C6.13962 -0.0536598 5.79164 -0.0536598 5.57702 0.160979L1.12662 4.61138C0.911981 4.826 0.911981 5.17398 1.1266 5.38862Z"
          fill="#959595"
        />
      </svg>
    </button>
    <div
      class="sc-dkzDqf eDaswd"
    >
      Content
    </div>
  </nav>
  <span
    data-focus-scope-end="true"
    hidden=""
  />
</DocumentFragment>
`;

exports[`SidebarNav toggles nav collapse on button click 2`] = `
<DocumentFragment>
  <span
    data-focus-scope-start="true"
    hidden=""
  />
  <nav
    aria-expanded="false"
    class="sc-bczRLJ fLqWsF sc-jSMfEi collapsed"
    data-testid="sidebarnav"
  >
    <button
      aria-label="Expand navigation"
      class="sc-eCYdqJ eKnMif collapsed"
      data-testid="sidebarnav-toggle"
    >
      <svg
        fill="none"
        height="10"
        viewBox="0 0 8 10"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.87337 5.38862L2.42295 9.83902C2.20831 10.0537 1.86033 10.0537 1.64571 9.83902L1.12664 9.31995C0.912369 9.10568 0.911957 8.75841 1.12573 8.54363L4.65276 4.99999L1.12573 1.45637C0.911957 1.24159 0.912369 0.89432 1.12664 0.680047L1.64571 0.160979C1.86035 -0.0536598 2.20833 -0.0536598 2.42295 0.160979L6.87335 4.61138C7.08799 4.826 7.08799 5.17398 6.87337 5.38862Z"
          fill="#959595"
        />
      </svg>
    </button>
    <div
      class="sc-dkzDqf eDaswd"
    >
      Content
    </div>
  </nav>
  <span
    data-focus-scope-end="true"
    hidden=""
  />
</DocumentFragment>
`;
